File: deploy.rb
require 'sinatra'
require 'json'
require 'logger'

$LOG = Logger.new('git-deploy.log',0,100*1024*1024)

post '/payload', :agent => /GitHub-Hookshot/ do
  request.body.rewind
  payload_body = request.body.read
  verify_signature(payload_body)
  commands = ['git status','git pull origin master','git submodule sync','git submodule update','git submodule status']
  commands.each do |i|
  	system i
  end
  
  $LOG.info "Git deploy ran without critical error."
  return "Git deploy ran without critical error."
end

def verify_signature(payload_body)
  signature = 'sha1=' + OpenSSL::HMAC.hexdigest(OpenSSL::Digest.new('sha1'), ENV['GIT_TOKEN'], payload_body)
  return halt 500, "Signatures didn't match!" unless Rack::Utils.secure_compare(signature, request.env['HTTP_X_HUB_SIGNATURE'])
end

File: deploy.php
<?php
	/**
	 * GIT DEPLOYMENT SCRIPT
	 *
	 * Used for automatically deploying websites via github or bitbucket, more deets here:
	 *
	 *		https://gist.github.com/1809044
	 */

	$agent=$_SERVER['HTTP_USER_AGENT'];
	$signature=$_SERVER['HTTP_X_HUB_SIGNATURE'];
	$body=@file_get_contents('php://input');

	// The commands
	$commands = array(
		'git pull origin master',
		'git status',
		'git submodule sync',
		'git submodule update',
		'git submodule status',
	);

	base64_encode($agent);
	base64_encode($signature);
	if (strpos($agent,'GitHub-Hookshot') !== false){
		if (hash_equals($signature, verify_request())){
			error_log("Success!");
		}else{
			error_log("Signature is invalid.");
		}
	}else{
		error_log('Request header is invalid.');
	}


	function verify_request(){
		$message = $GLOBALS['body'];
		$key     = "6217e99d55719fcad18aeed6f19fb9bcee225d1d";
	    $hash    = hash_hmac("sha1", $message, $key);
	    return $hash;
	}

	// Run the commands
	foreach($commands AS $command){
		// Run it
		$tmp = shell_exec($command);
	}

?>